qiskit/pulse/builder.py

```diff
@@ -895,6 +895,98 @@ def align_sequential() -> ContextManager[None]:
     """
 
 
+# pylint: disable=unused-argument
+@_transform_context(transforms.align_equispaced)
+def align_equispaced(duration: int) -> ContextManager[None]:
+    """Equispaced alignment pulse scheduling context.
+
+    Pulse instructions within this context are scheduled with the same interval.
+    If the total free induction time cannot be divided by the number of sub-schedules
+    within the context, the modulo is prepended and appended to the returned schedule.
+    Delay instruction is automatically inserted in between pulses.
+
+    This context may be convenient to write a schedule of the periodical dynamic decoupling or
+    the Hahn echo sequence.
+
+    Examples:
+
+    .. jupyter-execute::
+
+        from qiskit import pulse
+
+        d0 = pulse.DriveChannel(0)
+        x90 = pulse.Gaussian(10, 0.1, 3)
+        x180 = pulse.Gaussian(10, 0.2, 3)
+
+        with pulse.build() as hahn_echo:
+            with pulse.align_equispaced(duration=100):
+                pulse.play(x90, d0)
+                pulse.play(x180, d0)
+                pulse.play(x90, d0)
+
+        hahn_echo.draw()
+
+    Args:
+        duration: Duration of this context. This should be larger than the schedule duration.
+
+    Notes:
+        The scheduling is performed for sub-schedules within the context rather than
+        channel-wise. If you want to apply the equispaced context for each channel,
+        you need to apply the context independently to channels.
+    """
+
+
+# pylint: disable=unused-argument
+@_transform_context(transforms.align_numerical)
+def align_numerical(duration: int,
+                    position: Callable[[int], float]) -> ContextManager[None]:
+    """Numerically-defined alignment pulse scheduling context.
+
+    Pulse instructions within this context are scheduled at the location specified by
+    arbitrary callback function `position` that takes integer index and returns
```

Comment:
This should probably also accept the instruction to be scheduled.

Reply:
bump on this. Also why is the fractional position returned? Why not just return an explicit time?

Reply:
It looks like the convention of DD papers.